test/RestrictedNORI-transfer.test.ts

```diff
@@ -0,0 +1,531 @@
+          await batchMintAndListRemovalsForSale({
+            testSetup,
+            removalDataToList: [removalDataToList[0]],
+          });
+        const { listedRemovalIds: listedRemovalIds2, projectId: projectId2 } =
+          await batchMintAndListRemovalsForSale({
+            testSetup,
+            projectId: 999_999_999,
+            removalDataToList: [removalDataToList[1]],
+          });
+        const { supplier, investor1 } = hre.namedAccounts;
+        const restrictedAmounts = removalDataToList.map(
+          (removalData) => removalData.amount
+        );
+        await restrictRemovalProceeds({
+          testSetup,
+          removalIds: [listedRemovalIds1[0], listedRemovalIds2[0]],
+          removalAmountsToRestrict: restrictedAmounts,
+        });
+        const supplierScheduleDetailsBeforeTransfer =
+          await rNori.batchGetScheduleDetailsForAccount(supplier);
+
+        const expectedScheduleDetailsBeforeTransfer = [
+          {
+            tokenHolder: supplier,
+            scheduleTokenId: projectId1,
+            balance: BigNumber.from(restrictedAmounts[0]),
+          },
+          {
+            tokenHolder: supplier,
+            scheduleTokenId: projectId2,
+            balance: BigNumber.from(restrictedAmounts[1]),
+          },
+        ];
+        for (const [
+          index,
+          scheduleDetail,
+        ] of supplierScheduleDetailsBeforeTransfer.entries()) {
+          compareScheduleDetailForAddressStructs(
+            scheduleDetail,
+            expectedScheduleDetailsBeforeTransfer[index]
+          );
+        }
+
+        await rNori
+          .connect(hre.namedSigners.supplier)
+          .safeBatchTransferFrom(
+            supplier,
+            investor1,
+            [projectId1, projectId2],
+            [restrictedAmounts[0], restrictedAmounts[0]],
+            '0x'
+          );
+        const supplierScheduleDetailsAfterTransfer =
+          await rNori.batchGetScheduleDetailsForAccount(supplier);
+        const investor1ScheduleDetailsAfterTransfer =
+          await rNori.batchGetScheduleDetailsForAccount(investor1);
+
+        const expectedSupplierScheduleDetailsAfterTransfer = [
+          {
+            tokenHolder: supplier,
+            scheduleTokenId: projectId1,
+            balance: BigNumber.from(0),
+          },
+          {
+            tokenHolder: supplier,
+            scheduleTokenId: projectId2,
+            balance: BigNumber.from(0),
+          },
+        ];
+        const expectedInvestor1ScheduleDetailsAfterTransfer = [
+          {
+            tokenHolder: investor1,
+            scheduleTokenId: projectId1,
+            balance: BigNumber.from(restrictedAmounts[0]),
+          },
+          {
+            tokenHolder: investor1,
+            scheduleTokenId: projectId2,
+            balance: BigNumber.from(restrictedAmounts[1]),
+          },
+        ];
+
+        for (const [
```

Comment:
not in this pr, but this definitely needs a better abstraction in the future. OZ uses something like [this](https://github.com/OpenZeppelin/openzeppelin-test-helpers/blob/master/src/balance.js) to track balance changes and I think we could use the idea to create a DataTracker